src/targets/gpu/mlir.cpp

```diff
@@ -364,17 +388,41 @@ struct mlir_program
             lens.size(), lens.data(), strides.data(), make_type(s.type()));
     }
 
+    MlirType make_mlir_shaped(instruction_ref ins) const
+    {
+        shape s = ins->get_shape();
+        if(ins->name() == "@return")
+        {
+            assert(ins->inputs().size() == 1);
+            s = ins->inputs().front()->get_shape();
+        }
+        else if(instruction::get_output_alias(ins)->name() == "unpack_fp4")
+        {
+            s = s.with_type(shape::fp4x2_type);
+        }
+        else if(ins->get_shape().type() == shape::fp4x2_type)
+        {
+            auto new_lens = s.lens();
+            int fast_axis =
+                std::min_element(s.strides().cbegin(), s.strides().cend()) - s.strides().cbegin();
+            new_lens[fast_axis] *= 2;
+            s = s.with_lens(new_lens);
```

Comment:
This wont be correct if the shape is non-packed. You need to double the strides for everything except the fast axis and construct a new shape.

Reply:
Technically the shape should always be a packed one because the `pack_fp4` and `unpack_fp4` only make packed shapes. I'll make the change though.